Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,616 @@
+  ///   * `/Appliations/Foo.app/bar.txt => txt`
+  ///   * `/tmp/foo.tar.gz              => gz`
+  ///   * `/tmp/.hidden                 => nil`
+  ///   * `/tmp/.hidden.                => ""`
+  ///   * `/tmp/..                      => nil`
+  ///
+  /// Example:
+  ///
+  ///     var path = "/tmp/file"
+  ///     path.extension = ".txt" // path is "/tmp/file.txt"
+  ///     path.extension = ".o"   // path is "/tmp/file.o"
+  ///     path.extension = nil    // path is "/tmp/file"
+  ///     path.extension = ""     // path is "/tmp/file."
+  ///
+  public var `extension`: String? {
+    get { lastComponent?.extension }
+    set {
+      defer { _invariantCheck() }
+      guard let base = lastComponent, base.kind == .regular else { return }
+
+      let suffix: SystemString
+      if let ext = newValue {
+        suffix = _makeExtension(ext)
+      } else {
+        suffix = SystemString()
+      }
+
+      let extRange = (
+        base._extensionIndex() ?? base._slice.endIndex
+      ) ..< base._slice.endIndex
+
+      _storage.replaceSubrange(extRange, with: suffix)
+    }
+  }
+
+  /// The non-extension portion of the file or directory last component.
+  ///
+  /// Returns `nil` if `lastComponent` is `nil`
+  ///
+  ///   * `/tmp/foo.txt                 => foo`
+  ///   * `/Appliations/Foo.app/        => Foo`
+  ///   * `/Appliations/Foo.app/bar.txt => bar`
+  ///   * `/tmp/.hidden                 => .hidden`
+  ///   * `/tmp/..                      => ..`
+  ///   * `/                            => nil`
+  public var stem: String? { lastComponent?.stem }
+
+}
+
+extension FilePath {
+  /// Whether the path is in lexical-normal form, that is `.` and `..`
+  /// components have been collapsed lexically (i.e. without following
+  /// symlinks).
+  ///
+  /// Examples:
+  /// * `"/usr/local/bin".isLexicallyNormal == true`
+  /// * `"../local/bin".isLexicallyNormal   == true`
+  /// * `"local/bin/..".isLexicallyNormal   == false`
+  public var isLexicallyNormal: Bool {
+    // `..` components are permitted at the front of a
+    // relative path, otherwise there should be no special directories
+    //
+    // FIXME: Windows `C:..\foo\bar` should probably be lexically normal, but
+    // `\..\foo\bar` should not.
+    components.drop(
+      while: { root == nil && $0.kind == .parentDirectory }
+    ).allSatisfy { $0.kind == .regular }
+  }
+
+  /// Collapse `.` and `..` components lexically (i.e. without following
+  /// symlinks).
+  ///
+  /// Examples:
+  /// * `/usr/./local/bin/.. => /usr/local`
+  /// * `/../usr/local/bin   => /usr/local/bin`
+  /// * `../usr/local/../bin => ../usr/bin`
+  public mutating func lexicallyNormalize() {
+    defer { _invariantCheck() }
+    _normalizeSpecialDirectories()
+  }
+
+  /// Returns a copy of `self` in lexical-normal form, that is `.` and `..`
+  /// components have been collapsed lexically (i.e. without following
+  /// symlinks). See `lexicallyNormalize`
+  public var lexicallyNormal: FilePath {
```

Comment:
It's a little surprising to me that this is a computed property `lexicallyNormal` rather than a method `lexicallyNormalized()` because it has a mutating counterpart `lexicallyNormalize()`. (I think `isLexicallyNormal` is fine).